calico/openstack/test/test_plugin_etcd.py

```diff
@@ -495,3 +496,78 @@ def test_check_segment_for_agent(self):
             {mech_calico.api.NETWORK_TYPE: 'vlan'},
             mech_calico.constants.AGENT_TYPE_DHCP
         ))
+
+    def test_neutron_rule_to_etcd_rule_icmp(self):
+        # No type/code specified
+        self.assertNeutronToEtcd(neutron_rule({
+            "ethertype": "IPv4",
+            "protocol": "icmp",
+        }), {
+            'ip_version': 4,
+            'protocol': 'icmp',
+            'src_net': '0.0.0.0/0',
+        })
+        # Type/code wildcarded, same as above.
+        self.assertNeutronToEtcd(neutron_rule({
+            "ethertype": "IPv4",
+            "protocol": "icmp",
+            "port_range_min": -1,
+            "port_range_max": -1,
+        }), {
+            'ip_version': 4,
+            'protocol': 'icmp',
+            'src_net': '0.0.0.0/0',
+        })
+        # Type and code.
+        self.assertNeutronToEtcd(neutron_rule({
+            "ethertype": "IPv4",
+            "protocol": "icmp",
+            "port_range_min": 123,
+            "port_range_max": 100,
+        }), {
+            'ip_version': 4,
+            'protocol': 'icmp',
+            'src_net': '0.0.0.0/0',
+            'icmp_type': 123,
+            'icmp_code': 100,
+        })
+        # Type and code, IPv6.
+        self.assertNeutronToEtcd(neutron_rule({
+            "ethertype": "IPv6",
+            "protocol": "icmp",
+            "port_range_min": 123,
+            "port_range_max": 100,
+        }), {
+            'ip_version': 6,
+            'protocol': 'icmpv6',
+            'src_net': '::/0',
+            'icmp_type': 123,
+            'icmp_code': 100,
+        })
+
+    def assertNeutronToEtcd(self, neutron_rule, exp_etcd_rule):
```

Comment:
minor point: found it a little confusing at first that neutron_rule is trumping the function with the same name.